### ISSUE TITLE
Consider tweaking evil regex detection

### ISSUE THREAD

**Issue by AlexTMjugador** [blargbot/blargbot#159]
In my humble opinion, the current state of evil regex detection may cripple useful functionality of the Javascript regular expression engine to detect certain complex (but common) patterns, without necessarily compromising that functionality for decreased resource usage of the bot on its servers.

For instance, the following Javascript regex for matching e-mail addresses according to RFC 5322 (taken from http://emailregex.com/) fails to compile on Blargbot because it is considered evil:

`{regexmatch;myemail@server.com;/^(([^<>()\[\]\\.,{semi}:\s@"]+(\.[^<>()\[\]\\.,{semi}:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/}`

Another regex that failed to compile on Blargbot because this reason was [this one I came up with to validate URIs according to RFC 3986](https://regex101.com/r/jA6jJ7/5) (it might indeed be evil, but I really hope it is not. Others I found on Internet can even get more complex, because they validate the authority field too and are even more standards-conforming). As you can test on the interactive Javascript regex page I linked to, its performance seems _good enough_ for arbitrary input strings (I've tested it with several pretty long Google Image result data URIs and the matches were returned in less than 3 ms). For Blargbot I used a shortened version that recognizes less schemes, in order to stay below the 2000 character limit, without any more changes.

It's probable you're wondering why I want such complex regex matching for a Discord server. The answer is that currently I'm coding an everything autoresponse in BBTag to give users warnings automatically if they use caps so much, and I wish to be able to ignore their message if it is just an URI, because it makes little sense to warn users for caps used in a link path or query strings (the e-mail regex was just another example I thought it would be worth mentioning; I have no plans on matching e-mail addresses right now). Of course, I probably could have done an easier (but more error-prone) regex that works only with common URI formats as those used on HTTP(S), and it would probably be _good enough_, but stumbling upon this gave me an improvement suggestion for Blargbot anyway (_good enough_ doesn't mean _not able to be improved_).

I came up with serveral proposals to address this issue (you might find them unapplicable or simply plain bad ideas, though):

- Clearly warn in the documentation that evil regexes are checked for and won't compile, and hint what kind of regexes could be considered evil. _This doesn't fix the issue itself, but at least it would stop programmers like me spending energy to build a expectedly-very-good regex to only find it fail on Blargbot._

- Consider removing [this apparently spurious check](https://github.com/blargbot/blargbot/blob/362896dd83b922d9023a35ec7fdc063118ffbbc1/src/utils/generic.js#L1385) that marks as unsafe any regex which contains a capture group with the `+` or `*` qualifiers inside, being the capture group appended by the `+` or `*` qualifiers. _At best, I think that the usage of the `safe-regex` library that follows should be enough to detect evil regexes with _good enough_ accuracy and false positives. And this check doesn't take into account alternative qualifiers with similar effects such as `{1,}`. A library should be able to detect these trivial cases of evil regexes better, with less false positives, and if not then I'd consider switching to other library._

- Study the applicability of more efficient ways to compile and execute regexes, so evil regexes could be dealt with in a more graceful manner than simply refusing to compile them. _For instance, the regex matching algorithm could be run in a separate Javascript thread, which is aborted if it takes too long to run. Or improve the regex engine itself, by ensuring it runs in O(n) time for the most part (this is possible at least theoretically, I don't know if it is feasible to implement on Javascript)._

Some interesting reference I've found about the subject:

https://en.wikipedia.org/wiki/ReDoS#Description
https://www.reddit.com/r/programming/comments/gua22/regex_denial_of_service_redos/
https://swtch.com/~rsc/regexp/regexp1.html

I hope that was not too long to read 😄

**Comment by Ratismal** [comment]
I'm not exactly sure what the rationale behind using `/\(.*[*+].*\)[+*]/` in addition to safe-regex was. I can only assume that it was in response to something getting past safe-regex. I have no idea what this could possibly be since safe-regex should cover it, so I can try removing it.

It should be noted, however, that both of your expressions are deemed "not safe" by safe-regex regardless. There is simple not a good way to accurately detect evil regex while minimizing false positives, and I would rather be safe than sorry.

Generally, the regex that you'd use in your project would be created by you. You'd have control over whether it's safe or a DDoS waiting to happen. However, when dealing with arbitrary regex that potentially malicious users supply, it is very important to be cautious. I hope you realize the risk that this entails, and why the regex checking is so strict.

In response to your suggestions:
1. I can definitely add notes about unsafe regex
2. I can remove that specific check, but it will not solve your issue
3. Spawning a new thread for regex checks is ridiculous, and I will not do it. Rebuilding or modifying the regex engine is also not going to happen.

Now onto your personal situation.

To put it bluntly, [the second regex you provided](https://regex101.com/r/jA6jJ7/5) is pretty bad. I feel like I died a little looking at it. When your regex is ***2776*** characters long, it's time to take a step back and ask "is this really the best way to do this?" For starters, all of your hardcoded (🤢) protocols could be effectively replaced with `[\w.]+`

Next, on http://emailregex.com I found this safe regex: `/^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$/`

In all my time working with BBTag and regex in general, I have never found an expression which couldn't be simplified into something safe. This is no exception.

Please let me know if you have any additional thoughts on this matter. Thank you.

**Comment by AlexTMjugador** [comment]
Thanks for your timely and insightful response, I appreciate it.

I understand that allowing users to run their own regular expressions on arbitrary strings is prone to availability concerns for Blargbot, due to the risk that, intendedly or not, a regular expression runs for a lot of time, becoming a resource hog for which affects the rest of guilds in the same shard. So I think we agree that it is better to err on the safe side, even if it implies some loss of functionality.

About the regular expressions I proposed, I definitely agree that the second one is pretty bad. I looked several times at it and also wondered if there would be a simpler expression that works equally _good enough_ for the messages that could be sent on my server. Concerning the hardcoding of scheme names (or protocols), I understand that it is a brittle practice and bloats the regex, but I do not want to consider text such as `DSFGSFIOUHGFLH:IUGH` as an URI (something a user could write if he or she slams the keyboard properly, or even exploit on purpose). It is obvious such URIs aren't meant to be, even though they are technically valid according to the standard. So I took a safe approach and decided to only consider valid URIs those that use one of the recognized schemes, appropiate character sets for their components and a non-empty path (except if an authority component is present). The downsides I see for this approach is the added complexity to the regex and the possibility of a legit URI with an unrecognized scheme (or with non-canonical capitalization) not being detected, but as I'm the only mantainer of the code and the usage of strange schemes is a extreme edge case, I think it's fair to assume it is worth the effort because it's more likely a server member uses such legit URIs to get around the caps detection rather than by accident.

Luckily for me, and I hope for Blargbot too, after revising my requirements I ended up with a simpler regex that worked _even better_ for strange URIs, no qualifiers applying to capture groups (they are only used inside the capture groups) and a shorter list of hardcoded schemes which fits under 2000 characters and succeeds to compile. Although it is still visually complex, its complexity comes from alternatives and character classes, things that shouldn't kick execution times to exponential functions. A test with random text input showed short and consistent response times. I think it is a good balance between URI validation and code performance for my use case, because things like `http://?/page.htm`, `ftp://` or `HEY:YOU` are not considered an URI, while correctly assesing as valid URIs the strings `http://website.com/?myquery`, `blob:/`, `data:mydata/subpath` or `file:///C:/myfile.txt`.

As for the proposals, I'm definitely looking forward for the first two to be applied 😃. Maybe I can help on that by submitting a pull request with both suggestions, or one pull request for each suggestion, as you prefer. I understand this issue is not of enough priority to stop what Blargbot developers could be doing currently.
